### PR TITLE
fix(mdns): Don't expire mDNS nodes on connection close

### LINKED ISSUE #3309: mDNS records expire when connection times out
## Summary

mDNS records are expiring when an unrelated connection timeout with said peer is reached (eg. the gossipsub default of 120 seconds, or the floodsub default of… 10 seconds?). The problem is that connections may close for any unrelated reason. This does not render the peer unreachable. It is completely valid to reopen a connection to the peer again in many cases.

From the docs of `mdns::Event`:

> Each discovered record has a time-to-live. When this TTL expires and the address hasn’t been refreshed, we remove it from the list and emit it as an Expired event.`

The current behavior differs from the documentation and will expire nodes for any connection related timeout that has nothing to do with mDNS.

The [examples](https://github.com/libp2p/rust-libp2p/blob/master/examples/chat-tokio.rs#L146) are suggesting that an expired node is not reachable anymore and remove them from eg. gossipsub/floodsub. But this is not true if the connection has stopped due to a timeout and the peer remains reachable / the TTL has not yet expired.

## Expected behaviour

mDNS nodes only expire when their DNS TTL is reached and unrelated connection issues and timeouts will not expire mDNS nodes.

## Actual behaviour

mDNS nodes expire when a related connection times out. This makes floodsub-based chats completely unusable and gossipchat-based chats will expire after 120 seconds.

Debug output of the connection timeout in example `chat-tokio`. I added a few println! statements to the mDNS events. The peer is expired 10 seconds after discovery.

<details><summary>Debug Output</summary>
<p>

```
[2023-01-06T01:20:13Z INFO  libp2p_mdns::behaviour] discovered: 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp /ip4/192.168.122.1/tcp/33273
[2023-01-06T01:20:13Z INFO  libp2p_mdns::behaviour] discovered: 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp /ip4/192.168.42.60/tcp/33273
Discovered peer 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp
Discovered peer 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp
[2023-01-06T01:20:13Z DEBUG libp2p_tcp] dialing 192.168.122.1:33273
[2023-01-06T01:20:13Z DEBUG libp2p_tcp] dialing 192.168.42.60:33273
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG libp2p_tcp] New listen address: /ip4/192.168.42.60/tcp/41537
[2023-01-06T01:20:13Z DEBUG libp2p_swarm] Listener ListenerId(3635681165069297728); New address: "/ip4/192.168.42.60/tcp/41537"
Listening on "/ip4/192.168.42.60/tcp/41537"
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /noise
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /noise
[2023-01-06T01:20:13Z DEBUG libp2p_tcp] New listen address: /ip4/192.168.122.1/tcp/41537
[2023-01-06T01:20:13Z DEBUG libp2p_swarm] Listener ListenerId(3635681165069297728); New address: "/ip4/192.168.122.1/tcp/41537"
Listening on "/ip4/192.168.122.1/tcp/41537"
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG libp2p_tcp] Incoming connection from /ip4/192.168.42.60/tcp/59240 at /ip4/192.168.42.60/tcp/41537
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /noise
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /noise
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /noise
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /noise
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /mplex/6.7.0
[2023-01-06T01:20:13Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /mplex/6.7.0
[2023-01-06T01:20:13Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /mplex/6.7.0
[2023-01-06T01:20:13Z DEBUG libp2p_mplex::io] New multiplexed connection: f36a73bdbc5b8d46
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp") Listener { local_addr: "/ip4/192.168.42.60/tcp/41537", send_back_addr: "/ip4/192.168.42.60/tcp/59240" }; Total (peer): 1. Total non-banned (peer): 1
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /mplex/6.7.0
[2023-01-06T01:20:13Z DEBUG libp2p_mplex::io] f36a73bdbc5b8d46: New outbound substream: (0/initiator) (total 1)
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /floodsub/1.0.0
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /mplex/6.7.0
[2023-01-06T01:20:13Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /floodsub/1.0.0
[2023-01-06T01:20:13Z DEBUG libp2p_mplex::io] f36a73bdbc5b8d46: Closed substream (0/initiator) (half-close)
[2023-01-06T01:20:13Z DEBUG libp2p_mplex::io] New multiplexed connection: 5b1d18a264ab973a
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG libp2p_mplex::io] 5b1d18a264ab973a: New inbound substream: (0/receiver) (total 1)
[2023-01-06T01:20:13Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp") Dialer { address: "/ip4/192.168.122.1/tcp/33273/p2p/12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp", role_override: Dialer }; Total (peer): 2. Total non-banned (peer): 2
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /floodsub/1.0.0
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /floodsub/1.0.0
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:13Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:13Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:23Z DEBUG libp2p_swarm] Connection closed with error KeepAliveTimeout: Connected { endpoint: Listener { local_addr: "/ip4/192.168.42.60/tcp/41537", send_back_addr: "/ip4/192.168.42.60/tcp/59240" }, peer_id: PeerId("12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp") }; Total (peer): 1.
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:23Z DEBUG libp2p_swarm] Connection closed with error KeepAliveTimeout: Connected { endpoint: Dialer { address: "/ip4/192.168.122.1/tcp/33273/p2p/12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp", role_override: Dialer }, peer_id: PeerId("12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp") }; Total (peer): 0.
[2023-01-06T01:20:23Z DEBUG libp2p_tcp] dialing 192.168.122.1:33273
[2023-01-06T01:20:23Z DEBUG libp2p_tcp] dialing 192.168.42.60:33273
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] reading incoming messages
[2023-01-06T01:20:23Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] handling requests
[2023-01-06T01:20:23Z DEBUG netlink_proto::connection] sending messages
[2023-01-06T01:20:23Z INFO  libp2p_mdns::behaviour] expired: 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp /ip4/192.168.122.1/tcp/33273
[2023-01-06T01:20:23Z INFO  libp2p_mdns::behaviour] expired: 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp /ip4/192.168.42.60/tcp/33273
Expired peer 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp
[2023-01-06T01:20:23Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /noise
Removing peer 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp from floodsub
[2023-01-06T01:20:23Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /noise
Expired peer 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp
Removing peer 12D3KooWQ9fYo4U7eckzGsftxA69bC2iQUg5jjSwxVvQNCgiqMKp from floodsub
```
</p>
</details>

## Version

- libp2p version (version number, commit, or branch): master

## Would you like to work on fixing this bug?

Maybe

### PR DESCRIPTION
**Pull request by felinira**
## Description

mDNS records should not be expiring when an unrelated connection timeout with said peer is reached.

Fixes #3309.

## Links to any relevant issues

#3309

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/mdns/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
 - Update to `libp2p-swarm` `v0.42.0`.
 
+- Don't expire mDNS records when the last connection was closed. mDNS records will only be expired when the TTL is reached and the DNS record is no longer valid.
+
 [PR 3153]: https://github.com/libp2p/rust-libp2p/pull/3153
```
Comment (thomaseizinger) [inline protocols/mdns/CHANGELOG.md]:
```suggestion
- Don't expire mDNS records when the last connection was closed.
  mDNS records will only be expired when the TTL is reached and the DNS record is no longer valid.
  See [PR 3367].

[PR 3153]: https://github.com/libp2p/rust-libp2p/pull/3153
[PR 3367]: https://github.com/libp2p/rust-libp2p/pull/3367
```

**Review by thomaseizinger — COMMENTED** [review]
Thanks!

Can we add a test for this? It should be reasonably straight-forward to:

1. Boot up two nodes listening on TCP
1. Have them discover via mDNS
1. Connect to them
1. Disconnect them
1. Assert that we can still connect to them by _only_ specifying the `PeerId`, thus ensuring that the mDNS behaviour still uses the previously discovered record

**Comment by felinira** [comment]
I implemented the test only for the async-std backend. Should this be duplicated for tokio or possibly live somewhere else?

**Comment by thomaseizinger** [comment]
> I implemented the test only for the async-std backend. Should this be duplicated for tokio or possibly live somewhere else?

That is fine, thank you :)

**Review by thomaseizinger — APPROVED** [review]
There seems to be a typo in the test name, otherwise LGTM :)

**Comment by felinira** [comment]
The typo helped me realize I somehow committed a working copy of the test instead of the final result. This one should work though.

File: protocols/mdns/tests/use-async-std.rs [hunk 2]
```diff
@@ -126,3 +144,59 @@ async fn run_peer_expiration_test(config: Config) -> Result<(), Box<dyn Error>>
...
+            ev = a.select_next_some() => match ev {
+                SwarmEvent::Behaviour(Event::Discovered(peers)) => {
+                    for (peer, addr) in peers {
+                        if peer == *b.local_peer_id() && !dialed {
+                            // Connect to all addresses of b to 'expire' all of them
+                            a.dial(addr)?;
+                            dialed = true;
+                        }
+                    }
+                }
```
Comment (thomaseizinger) [inline protocols/mdns/tests/use-async-std.rs]: Would it be better to not iterate and just take the first one? Overriding the boolean state multiple times seems a bit odd.

File: protocols/mdns/tests/use-async-std.rs [hunk 3]
```diff
@@ -126,3 +144,59 @@ async fn run_peer_expiration_test(config: Config) -> Result<(), Box<dyn Error>>
         }
     }
 }
+
+async fn run_no_expiration_on_close_test(config: Config) -> Result<(), Box<dyn Error>> {
+    let mut a = create_swarm(config.clone()).await?;
+    let mut b = create_swarm(config).await?;
+
+    let mut closed = false;
+    let mut dialed = false;
```
Comment (thomaseizinger) [inline protocols/mdns/tests/use-async-std.rs]: Perhaps an enum would express this a bit better?

File: protocols/mdns/tests/use-async-std.rs (hunk 2 above)
Comment (felinira) [reply protocols/mdns/tests/use-async-std.rs]: The first one might not be b but someone else or even a. But you are right this was a bit convoluted.

File: protocols/mdns/tests/use-async-std.rs (hunk 3 above)
Comment (felinira) [reply protocols/mdns/tests/use-async-std.rs]: Replaced it with an enum

**Comment by felinira** [comment]
Ah, now this doesn't work either because it passes the without the MR again.

**Comment by felinira** [comment]
This works now. This test is a bit tricky because it might happen that the same peer is discovered multiple times and therefore it needs quite a bit of logic to determine the current state to make sure we are not introducing the address again at a later time.

**Review by thomaseizinger — APPROVED** [review]
One clippy failure, otherwise LGTM!

**Comment by thomaseizinger** [comment]
@mergifyio refresh

**Comment by thomaseizinger** [comment]
@mergifyio refresh

**Review by thomaseizinger — DISMISSED** [review]

**Review by thomaseizinger — APPROVED** [review]

**Review by mxinden — COMMENTED** [review]
:+1: thanks for the help here!

**Pull request merged**